Cleanup async execution for python RPC calls.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#53230 Cleanup async execution for python RPC calls.**
* #53207 Simplify async execution for script remote calls.
* #53204 Simplify async execution for script calls.

As part of https://github.com/pytorch/pytorch/issues/39351, cleaning
up the python call async execution.

Differential Revision: [D26800758](https://our.internmc.facebook.com/intern/diff/D26800758/)

pritamdamania committed:
Cleanup async execution for python RPC calls.

As part of https://github.com/pytorch/pytorch/issues/39351, cleaning
up the python call async execution.

Differential Revision: [D26800758](https://our.internmc.facebook.com/intern/diff/D26800758/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -209,39 +210,27 @@ void RequestCallbackImpl::processPythonCall(
-    markComplete(
-        std::move(PythonResp(std::move(*serializedPyObj))).toMessage());
+  try {
+    processPythonAsyncExecution(
```
nit: since its not necessarily async, rename it to `processPythonExecution`?

rohan-varma commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -358,52 +347,34 @@ void RequestCallbackImpl::processPythonRemoteCall(
+            const std::shared_ptr<JitFuture>& responseFuture) {
+          IValue py_ivalue = jit::toIValue(result, PyObjectType::get());
+
+          py::gil_scoped_release release;
```
Unrelated and does not need to be addressed in this diff, but feel like this could be challenging to reason about. We acquire GIL before call to `postProcessing`, and then release it here, and then it's re-acquired back in the original scope, I think. Should we at least have a DCHECK() here that the GIL is held?

pritamdamania committed:
Update on "Cleanup async execution for python RPC calls."


As part of https://github.com/pytorch/pytorch/issues/39351, cleaning
up the python call async execution.

Differential Revision: [D26800758](https://our.internmc.facebook.com/intern/diff/D26800758/)

[ghstack-poisoned]
